File: Algorithm/set_matrix_zeroes.py
class Solution:
    def setZeroes(self,matrix):
        row=len(matrix)
        col=len(matrix[0])
        row_flag=[False for x in range(row)]
        col_flag=[False for y in range(col)]
        for x in range(row):
            for y in range(col):
                if matrix[x][y]==0:
                    row_flag[x]=True
                    col_flag[y]=True
        for x in range(row):
            for y in range(col):
                if row_flag[x]:
                    matrix[x][y]=0
                elif col_flag[y]:
                    matrix[x][y]=0
        return matrix
    def setZeroes_1(self,matrix):
        row=len(matrix)
        col=len(matrix[0])
        row_flag=False
        col_flag=False
        for x in range(row):
            if matrix[x][0]==0:
                row_flag=True
                break
        for y in range(col):
            if matrix[0][y]==0:
                col_flag=True
                break
        for x in range(1,row):
            for y in range(1,col):
                if matrix[x][y]==0:
                    matrix[x][0]=0
                    matrix[0][y]=0
        for x in range(1,row):
            for y in range(1,col):
                if matrix[x][0]==0:
                    matrix[x][y]=0
                elif matrix[0][y]==0:
                    matrix[x][y]=0
        if row_flag:
            for x in range(row):
                matrix[x][0]=0
        if col_flag:
            for y in range(col):
                matrix[0][y]=0
        return matrix
                     



if __name__=='__main__':
    sol=Solution()
    matrix=[[1,2,3],[4,0,5],[6,7,8]]
    print sol.setZeroes(matrix)
    matrix=[[1,2,3],[4,0,5],[6,7,8]]
    print sol.setZeroes_1(matrix)

File: Algorithm/CPP/set_matrix_zeroes.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <algorithm>
# include<time.h>
using namespace std;

# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;
class Solution {
public:
	void setZeroes(vector<vector<int>>& matrix)
	{
		vector<bool> row(matrix.size()), col(matrix[0].size());
		for (auto i = 0; i != matrix.size(); ++i)
			for (auto j = 0; j != matrix[0].size(); ++j)
			{
				if (matrix[i][j] == 0)
				{
					row[i] = true;
					col[j] = true;
				}
			}
		for (auto i = 0; i != matrix.size(); ++i)
			for (auto j = 0; j != matrix[0].size(); ++j)
			{
				if (row[i] || col[j] )
				{
					matrix[i][j] = 0;
				}
			}

	}

};
int  main()
{
	Solution sol;
	vector<vector<int>> candidates = { { 2, 3, 6, },{ 7,0,1 }, { 4, 2, 0 }  };
	sol.setZeroes(candidates);
	for (auto k : candidates)
	{
		for (auto x : k)
		{
			cout << x << ", ";
		}

		cout << endl;
	}


	cin.get();
	return 0;
}
